dnnf/falsifier.py

```diff
@@ -1,304 +1,334 @@
+    outvars = get_concrete_inputs(model.prop.output_vars, kwargs)
+    for _, adv, _ in zip(success, advs, epsilons):
+        counter_example = normalizer(adv).cpu().detach().numpy()
+        if model.validate(counter_example, other_inputs=outvars):
+            logger.info("FOUND COUNTEREXAMPLE")
+            return counter_example
+        logger.debug("Counter example could not be validated")
+
+
+def tensorfuzz(model: FalsificationModel, n_steps=100, **kwargs):
+    import select
+    import shlex
+    import subprocess as sp
+    import tempfile
+
+    logger = logging.getLogger(__name__)
+
+    lb = model.input_lower_bound
+    ub = model.input_upper_bound
+
+    with tempfile.TemporaryDirectory() as tmpdir:
+        logger.debug("Using temporary directory: %s", tmpdir)
+
+        model_filename = f"{tmpdir}/model.onnx"
+        lb_filename = f"{tmpdir}/lb.npy"
+        ub_filename = f"{tmpdir}/ub.npy"
+
+        dummy_x = torch.from_numpy(lb)
+        torch.onnx.export(
+            model.model,
+            dummy_x,
+            model_filename,
+            input_names=["input"],
+            dynamic_axes={"input": [0]},
+        )
+        np.save(lb_filename, lb[0])
+        np.save(ub_filename, ub[0])
+
+        cmd = f"tensorfuzz.sh --model={model_filename} --lb={lb_filename} --ub={ub_filename} --label=0"
+        logger.debug("Running: %s", cmd)
+
+        proc = sp.Popen(
+            shlex.split(cmd), stdout=sp.PIPE, stderr=sp.PIPE, encoding="utf8"
+        )
+        assert proc.stderr is not None
+        assert proc.stdout is not None
+        stdout_lines: List[str] = []
+        stderr_lines: List[str] = []
+        while proc.poll() is None:
+            for (name, stream, lines) in [
+                ("STDOUT", proc.stdout, stdout_lines),
+                ("STDERR", proc.stderr, stderr_lines),
+            ]:
+                ready, _, _ = select.select([stream], [], [], 0)
+                if not ready:
+                    continue
+                line = stream.readline()
+                if line == "":
+                    continue
+                lines.append(line)
+                logger.debug(f"{{TENSORFUZZ ({name})}}: {line.strip()}")
+        for line in proc.stdout.readlines():
+            logger.debug(f"{{TENSORFUZZ (STDOUT)}}: {line.strip()}")
+        stdout_lines.extend(stdout_lines)
+        for line in proc.stderr.readlines():
+            logger.debug(f"{{TENSORFUZZ (STDERR)}}: {line.strip()}")
+        stderr_lines.extend(stderr_lines)
+        if "Fuzzing succeeded" in "\n".join(stderr_lines):
+            counter_example = np.load(f"{tmpdir}/cex.npy")[None].astype(
+                model.input_details[0].dtype
+            )
+            outvars = get_concrete_inputs(model.prop.output_vars, kwargs)
+            if model.validate(counter_example, other_inputs=outvars):
+                logger.info("FOUND COUNTEREXAMPLE")
+                return counter_example

```

Comment:
It looks like the line endings were changed in this file. Could you revert back to the original line endings so that we can get a better sense of the changes?